src/java.base/share/classes/com/sun/crypto/provider/GaloisCounterMode.java

```diff
@@ -624,6 +721,121 @@ int decryptFinal(byte[] in, int inOfs, int len,
         return len;
     }
 
+    // Note: In-place operations do not need an intermediary copy because
+    // the GHASH check was performed before the decryption.
+    int decryptFinal(ByteBuffer src, ByteBuffer dst)
+        throws IllegalBlockSizeException, AEADBadTagException,
+        ShortBufferException {
+        // Length of the input
+        ByteBuffer tag;
+        ByteBuffer ct = src.duplicate();
+        ct.mark();
+
+        ByteBuffer buffer = ((ibuffer == null || ibuffer.size() == 0) ? null :
+            ByteBuffer.wrap(ibuffer.toByteArray()));
+        int len;
+
+        if (ct.remaining() >= tagLenBytes) {
+            tag = src.duplicate();
+            tag.position(ct.limit() - tagLenBytes);
+            ct.limit(ct.limit() - tagLenBytes);
+            len = ct.remaining();
+            if (buffer != null) {
+                len += buffer.remaining();
+            }
+            ct.mark();
+        } else if (buffer != null && ct.remaining() < tagLenBytes) {
+            // It's unlikely the tag will be between the buffer and data
+            tag = ByteBuffer.allocate(tagLenBytes);
+            int limit = buffer.remaining() - (tagLenBytes - ct.remaining());
+            buffer.mark();
+            buffer.position(limit);
+            // Read from "new" limit to buffer's end
+            tag.put(buffer);
+            // reset buffer to data only
+            buffer.reset();
+            buffer.limit(limit);
+            tag.put(ct);
+            tag.flip();
+            // Limit is how much of the ibuffer has been chopped off.
+            len = buffer.remaining();
+        } else {
+            throw new AEADBadTagException("Input too short - need tag");
+        }
+
+        // do this check here can also catch the potential integer overflow
+        // scenario for the subsequent output buffer capacity check.
+        checkDataLength(0, len);
+
+        if ((ibuffer.size() + ct.remaining()) - tagLenBytes >
```

Comment:
ct is set a new limit (minus the tag length) on line 741. So this check seems incorrect?
How about using the 'len' value which seems to be the overall input size which should also be the expected output size?

Reply:
Yeah, I'd agree